internal/storage/dial.go

```diff
@@ -112,11 +112,42 @@ func registerOpenTelemetryDatabaseStats(db *pop.Connection) {
 	}
 }
 
+type CommitWithError struct {
+	Err error
+}
+
+func (e *CommitWithError) Error() string {
+	return e.Err.Error()
+}
+
+func (e *CommitWithError) Cause() error {
+	return e.Err
+}
+
+// NewCommitWithError creates an error that can be returned in a pop transaction
+// without rolling back the transaction. This should only be used in cases where
+// you want the transaction to commit but return an error message to the user.
+func NewCommitWithError(err error) *CommitWithError {
+	return &CommitWithError{Err: err}
+}
+
 func (c *Connection) Transaction(fn func(*Connection) error) error {
 	if c.TX == nil {
-		return c.Connection.Transaction(func(tx *pop.Connection) error {
-			return fn(&Connection{tx})
+		var returnErr error
+		terr := c.Connection.Transaction(func(tx *pop.Connection) error {
```

Comment:
/nit would it be possible to inline the `terr` check?

`if terr := ... ;terr!= nil{}`